packages/expo-image/src/ExpoImage.web.tsx

```diff
@@ -57,50 +62,172 @@ function getObjectPositionFromContentPosition(contentPosition?: ImageContentPosi
+) {
+  if (transition == null) {
+    return {
+      timing: ImageTransitionTiming.LINEAR,
+      duration: 0,
+      effect: ImageTransitionEffect.NONE,
+    };
+  }
+  if (typeof transition === 'number') {
+    return {
+      timing: ImageTransitionTiming.EASE_IN_OUT,
+      duration: transition,
+      effect: ImageTransitionEffect.CROSS_DISOLVE,
+    };
+  }
+  return {
+    timing: ImageTransitionTiming.EASE_IN_OUT,
+    duration: 1000,
+    ...transition,
+  };
+}
+
+const useTransition = (
+  transition: number | ImageTransition | null | undefined,
+  state: ImageState
+): Record<'placeholder' | 'image', Partial<React.CSSProperties>> => {
+  const { duration, timing, effect } = getTransitionObjectFromTransition(transition);
+  if (effect === ImageTransitionEffect.CROSS_DISOLVE) {
+    const commonStyles = {
+      transition: `opacity ${duration}ms`,
+      transitionTimingFunction: getCSSTiming(timing),
+    };
+    return {
+      image: {
+        opacity: state === 'loaded' ? '1' : '0',
+        ...commonStyles,
+      },
+      placeholder: {
+        opacity: state === 'loaded' ? '0' : '1',
+        ...commonStyles,
+      },
+    };
+  }
+  if (effect === ImageTransitionEffect.FLIP_FROM_TOP) {
+    const commonStyles = {
+      transition: `transform ${duration}ms`,
+      transformOrigin: 'top',
+      transitionTimingFunction: getCSSTiming(timing),
+    };
+    return {
+      placeholder: {
+        transform: `rotateX(${state !== 'loaded' ? '0' : '90deg'})`,
+        ...commonStyles,
+      },
+      image: {
+        transform: `rotateX(${state === 'loaded' ? '0' : '90deg'})`,
+        ...commonStyles,
+      },
+    };
+  }
+
+  return { placeholder: {}, image: {} };
 };
 
 export default function ExpoImage({
   source,
-  defaultSource,
+  placeholder,
   loadingIndicatorSource,
   contentPosition,
   onLoad,
+  transition,
   onLoadStart,
   onLoadEnd,
   onError,
   ...props
 }: ImageProps) {
   const { aspectRatio, backgroundColor, transform, borderColor, ...style } = props.style ?? {};
+  const [state, handlers] = useImageState(source);
+  const { placeholder: placeholderStyle, image: imageStyle } = useTransition(transition, state);
+
   const resolvedSources = ensureIsArray(source).map(resolveAssetSource);
+
   return (
-    <>
-      <picture
+    <div
+      style={{
+        aspectRatio: String(aspectRatio),
+        backgroundColor: backgroundColor?.toString(),
+        transform: transform?.toString(),
+        borderColor: borderColor?.toString(),
+        ...style,
+        overflow: 'hidden',
+        position: 'relative',
+      }}>
+      <img
+        src={ensureIsArray(placeholder).map(resolveAssetSource)?.[0]?.uri}
+        style={{
+          width: '100%',
+          height: '100%',
+          position: 'absolute',
+          left: 0,
+          right: 0,
+          objectFit: 'scale-down',
+          objectPosition: 'center',
+          ...placeholderStyle,
```

Comment:
In the future I think we can extract styles to separate functions, `render` would look cleaner 🙂

Reply:
I'll try and do that on final cleanup